tenant defined teletype template in callflow

sergey-safarov opened the pull request:
Added ability to define account level only teletype template (not template in system_config database) and then call this template from callflow.

sergey-safarov commented on applications/callflow/src/module/cf_customer_defined_notification.erl:
```diff
@@ -0,0 +1,166 @@
+            lager:debug("trying to send customer defined notification \"~s\" for call-id ~s", [TemplateID, kapps_call:call_id_direct(Call)]),
+            Emails = find_email_addresses(Call, kz_json:get_value(<<"recipients">>, Data, [])),
+            Props = props:filter_undefined(
+                      [{<<"From-User">>, kapps_call:from_user(Call)}
```
May need to add some other things that may be used in custom template?

sergey-safarov replied:
@bazemoreb could look this macros.
May be you need more.

sergey-safarov commented on applications/teletype/doc/customer_defined_notification.md:
```diff
@@ -0,0 +1,21 @@
+Possible macros
```
Could you suggest example how to format documentation for template macros.

jamesaimonetti replied:
One suggestions is to break them up by category:

## System
hostname | the system hostname
node | the system node

## Account
...

sergey-safarov replied:
Hello @jamesaimonetti 
Could you look updated macros description

sergey-safarov commented on applications/teletype/src/teletype_templates.erl:
```diff
@@ -175,11 +176,16 @@ render(TemplateId, Macros, DataJObj, 'true') ->
     lager:warning("no account id for template ~s, no template to process", [_TemplateId]),
     'undefined';
-templates_source(_TemplateId, ?KZ_CONFIG_DB) ->
-    ?KZ_CONFIG_DB;
```
This change may have side affect.
Is possible to cherry-pick this to `4.2`? Or this may be used only on `4.3`?

sergey-safarov commented on applications/teletype/src/teletype_templates.erl:
```diff
@@ -196,7 +202,7 @@ templates_source(TemplateId, AccountId, AccountId) ->
     case fetch_notification(TemplateId, AccountId) of
         {'ok', Template} ->
             templates_source_has_attachments(TemplateId, AccountId, AccountId, Template);
-        {'error', 'not_found'} -> ?KZ_CONFIG_DB;
```
This is not necessary, because `fetch_notification` also fetch `Reseller` and `system_config` notification templates.
If `fetch_notification` returned `not_found`, than notification template is not exist in `system_config`  too. And returning `system_config` will be false response.

k-anderson commented:
@sergey-safarov howdy! couple of initial thoughts:
1) customer_defined_notification is a bit of a verbose name, why not just 'cf_notification' or something more generic that support future growth?  For example, email is the currently support notification but later might include SMS or other notification methods.
2) I think it might be useful to be able to send a 'description' or 'comment' from the callflow that is then on the email.  Some data in the callflow node like:
"description": "You have received a call from the support line, please make sure you answer!"
3) For the email template rather than adding a new one, perhaps use the customer_update that already exists?

Again just some thoughts so open to discussions, thanks!

sergey-safarov commented:
@k-anderson very interest suggestion about SMS usage as notification.
Thing need create dedicated `text/sms` attachment for template with with very important  info only. And then user can select what is type is notifications is preferred (email, sms or both).
Also for this need define system wide and account wide policy/service. Like sms notification must be charged and may be disabled by default or disallowed at account level.

Think this task must be completed as dedicated PR.

About `description` filed in callflow. I will add this.
About `customer_update` also look sources.

bazemoreb commented:
Agreed @sergey-safarov - we are working on SMS piece on our side now (about 90% there) and then will add this on to our list to see if it makes sense.  For now we've just been "faking" it using the sms email services that carriers use.

bazemoreb commented:
Hi all - any update on getting this merged in?  Any further questions/concerns?

lazedo commented:
@bazemoreb the initial comments were not addressed. after that, code review will continue.

bazemoreb commented:
Ahh missed that, @lazedo .  @sergey-safarov could you address the items they had so we can get this merged in?

sergey-safarov commented on applications/callflow/doc/notification.md:
```diff
@@ -10,11 +10,14 @@ Validator for the customer defined notification callflow's data object
 --- | ----------- | ---- | ------- | -------- | -------------
+`comments` | Text message that need to include into notification | `string()` |   | `false` |  
+`notification_media.[]` |   | `string('email' | 'sms')` |   | `false` |  
+`notification_media` | What is media need to use for notification | `array(string('email' | 'sms'))` |   | `false` |  
```
Why validator wants two string definition  for `notification_media` rather one string?
How i can add description for  `notification_media.[]`?

sergey-safarov commented:
Hello @k-anderson, @lazedo 
Could you  look [this PR](https://github.com/2600hz/kazoo/pull/4979).
PR is updated:
1) Renamed callflow `customer_defined_notification' to 'cf_notification`;
2) Added filed `comments` and `notification_media` with enum values `sms`, `email`;
3) Removed unnecessary template `customer_defined_notification ` and for macros used `customer_update`.
PR is locally tested.

bazemoreb commented:
> @bazemoreb the initial comments were not addressed. after that, code review will continue.

Everything looking good here now. @lazedo ?

bazemoreb commented:
Hi team - checking to see if there are any further requests or revisions here prior to getting merged?

bazemoreb commented:
Bueller?

jamesaimonetti commented on applications/callflow/src/module/cf_notification.erl:
```diff
@@ -0,0 +1,170 @@
+handle(Data, Call) ->
+    case kz_json:get_ne_binary_value(<<"send_at">>, Data, <<"channel_destroy">>) of
+        <<"channel_destroy">> -> add_handler(Data, Call);
+        <<"callflow_exec">> -> send_notification(Call, [], Data);
```
You send an empty list here but the spec of `send_notification` expects a `kz_json:object()`. I suspect the spec should have it be `kz_json:json_term()` instead?

sergey-safarov replied:
I looked into [cf_exe:handle_channel_destroyed](https://github.com/2600hz/kazoo/blob/master/applications/callflow/src/cf_exe.erl#L850-L854) here is is passed `kz_json:object()`.
To clean code i can replace `[]` by `kz_json:new()`.

sergey-safarov commented on applications/callflow/src/module/cf_notification.erl:
```diff
@@ -0,0 +1,170 @@
+
+-spec send_notification(kapps_call:call(), kz_json:object(), kz_json:object()) -> 'ok'.
+send_notification(Call, Notify, Data) ->
+    case kz_json:get_value(<<"template_id">>, Data) of
```
@jamesaimonetti i expect this call will be broken after changing `send_notification` spec.
```erlang
-spec add_handler(kz_json:object(), kapps_call:call()) -> 'ok'.
add_handler(Data, Call) -> cf_exe:add_termination_handler(Call, {?MODULE, 'send_notification', [Data]}).
```
Could you make one more look?

jamesaimonetti commented:
I fixed a couple spelling issues and refactored a couple functions to be smaller but otherwise this looks worth getting into master and getting people trying it out. Thanks for everyone's patience!

bazemoreb commented:
Sweet!  Thanks James.

bazemoreb commented:
> Sweet! Thanks James.

Hey @jamesaimonetti - was looking to see if this one was merged, can you let me know or point in right direction?

sergey-safarov commented:
Hi Brandon @bazemoreb 
Merget at https://github.com/2600hz/kazoo/commit/d21cee323baab82199c29a671696e1c520a6b5d6
This is next 4.4 branch.
